common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,490 @@
+                                CancelCbaAndResetCurrentDialog();
+                                //TODO: create and show error dialog here.
+                            } catch (final BadResponseException e) {
+                                Logger.error(methodTag, "BadResponseException", e);
+                                CancelCbaAndResetCurrentDialog();
+                                //TODO: create and show error dialog here.
+                            }
+                        }
+                    });
+                }
+            }
+        };
+
+    }
+
+    /**
+     * Call current dialog's onCancelCba method and reset it back to null.
+     */
+    private void CancelCbaAndResetCurrentDialog() {
+        synchronized (smartcardDialogLock) {
+            if (mCurrentDialog != null) {
+                mCurrentDialog.onCancelCba();
+                mCurrentDialog = null;
+            }
+        }
+    }
+
+    /**
+     * Checks to see if PIN for smartcard is correct.
+     * If so, proceed to attempt authentication.
+     * Otherwise, handle the incorrect PIN based on how many PIN attempts are remaining.
+     * @param pin char array containing PIN attempt.
+     * @param certDetails YubiKitCertDetails of the selected certificate from the SmartcardCertPickerDialog.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @param piv A PivSession created from a SmartCardConnection that can interact with certificates located in the PIV slots on the YubiKey.
+     * @throws IOException          in case of connection error
+     * @throws ApduException        in case of an error response from the YubiKey
+     * @throws BadResponseException in case of incorrect YubiKey response
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void verifySmartcardPin(@NonNull final char[] pin, @NonNull final YubiKitCertDetails certDetails, @NonNull final ClientCertRequest request, @NonNull final PivSession piv) throws IOException, ApduException, BadResponseException {
+        final String methodTag = TAG + ":verifySmartcardPin";
+        //Call YubiKit method to verify PIN.
+        try {
+            piv.verifyPin(pin);
+            //If pin is successfully verified, we will get the certificate and perform the rest of the logic for authentication.
+            X509Certificate cert = piv.getCertificate(certDetails.getSlot());
+            //TODO: do a confirmation that the cert retrieved above is the same cert that was picked earlier by comparing with YubiKitCertDetails
+            //TODO: Complete authentication using cert and YubiKit sdk.
+            //TODO: Delete test code below between START and END comments (should be replaced with actual authentication)
+            //NOTE: below is for testing. This would get replaced by the logic for actual authentication.
+            //TODO:START of code for testing
```

Comment:
This is going to be replaced in next few days and NOT ship? Or rather is this testing code going to ship?

Reply:
replaced in next few days and NOT ship (it has already been replaced in the next PR I'm working on).